graphicsmagick-sys/wrapper.h

```diff
@@ -1 +1,4 @@
 #include <wand/magick_wand.h>
+
+// help bindgen automatically generate FFI binding
+unsigned int MagickSetDepth(MagickWand *wand, const size_t depth);
```

Comment:
Not maintainer, but I'm curious why did you add that to this file.
In general, the `graphicsmagick-sys` should be automatically generated from `wand/magick_wand.h`.

Is that API only available on some version?
If so, you should put `MagickWand::set_depth` behind a feature flag.

Reply:
I think bindgen can't automatically generate FFI binding for this function without our help. There is no MagickSetDepth function declaration in GraphicsMagick's header files. There is only `#define MagickSetDepth GmMagickSetDepth` in `wand/wand_symbols.h` and the function definition in `wand/magick_wand.c`. 

I haven't found a better way to expose this function, it would be great if we find it.

Reply:
If `MagickSetDepth` is a macro which invokes `GmMagickSetDepth`, then simply declaring `MagickSetDepth` in the header just wouldn't work.

It would throw an error at link time because such function does not exist in the library.

You would have to invoke `GmMagickSetDepth` instead.

Reply:
If I'm not mistaken, this function is available from version 1.3.13 (December 24, 2011).

Reply:
> There is only #define MagickSetDepth GmMagickSetDepth in wand/wand_symbols.h and

A macro is not the same as function declaration, in that it does not expose a new API.

> the function definition in wand/magick_wand.c.

That's an implementation detail and cannot be relied on.
The symbol for that function can be removed when building the final library.

Reply:
> If `MagickSetDepth` is a macro which invokes `GmMagickSetDepth`, then simply declaring `MagickSetDepth` in the header just wouldn't work.
<br> It would throw an error at link time because such function does not exist in the library.

I have checked the proposed changes before creating the pull request and it works, there is no linking error.

> You would have to invoke GmMagickSetDepth instead.

According to the [documentation](http://www.graphicsmagick.org/wand/magick_wand.html#magicksetdepth), I think we would have to invoke MagickSetDepth, GmMagickSetDepth is an an implementation detail.

Reply:
On my system wtih `GraphicsMagick-1.3.38`:
```shell
$ objdump -T /usr/lib/libGraphicsMagickWand.so | fgrep SetDepth
000000000001aa90 g    DF .text	0000000000000070  Base        MagickSetDepth
```

Reply:
It works for your system, but it may not work for other systems or with future release.
Using `GmMagickSetDepth` here is the most robust solution AFAIK.

Reply:
How about add `wrapper.c` file and add the new function to call the `MagickSetDepth`, and then use [`cc`](https://crates.io/crates/cc) to generate the .a?

Just like:

```rust
pub extern "C" fn GraphicsmagickRsMagickSetDepth(...) {
    return MagickSetDepth(...);
}
```

And then the `MagickWand::set_depth` use `GraphicsmagickRsMagickSetDepth`.

Reply:
I think I don't fully understand the idea yet. Could you explain what we want to achieve with this?

Reply:
So in additional to the `wrapper.h`, you would also create `wrapper.c` which includes:

```
#include <stddef.h>

unsigned GraphicsmagickRsMagickSetDepth(MagickWand *wand, size_t depth) {
    return MagickSetDepth(wand, depth);
}
```

Then in the `wrapper.h`, includes:

```c
extern "C" unsigned GraphicsmagickRsMagickSetDepth(MagickWand *wand, size_t depth);
```

In addition to linking with graphicsmagick, use crate [`cc`](https://docs.rs/cc) to compile `wrapper.c` and link with it.

Reply:
I didn't understand why to do this. What are the advantages?

Reply:
> I didn't understand why to do this. What are the advantages?

We can't call the C macro `MagickSetDepth` from rust, but we can call it from C.

Since it is a macro provided by graphicsmagick, its implementation can change in the future release, such as calling a different function, or changed from a C macro to be an actual function.

Calling it directly from C will ensure that it would work in the future.

Reply:
It looks like `<wand/magick_wand.h>` forgot to export `MagickSetDepth`. I think it's OK now re-define it in `wrapper.h`.

Reply:
I agree with @jmjoy , but I would feel more comfortable if the upstream can confirm this and fix the issue.
But I'm ok with the current approach, so you don't have to foce yourself to deal with upstream if you don't have time @irkill

Reply:
OK, so I will merge it.